### PR TITLE
Operator batch normalization

### PR DESCRIPTION
**Pull request by MartinNoc**
This PR contains the BatchNormalizationOperator class implementation which is simply done using a matrix-vector multiplication via the GemmWrappedOperator. `batchnorm.py` is a "unit-test" for the implemented operator.
Besides, some changes had to be made
- newer pre-commit package versions
- `onnxruntime.InferenceSession` now requires the `providers` parameter

`generate_mnist_calibration_data.py` is just a convenient utility script.

**UPDATE:**
BatchNormalizationOperator is now implemented in an efficient way as a vector-vector multiplication plus bias addition, i.e., a pointwise multiplication of the CKKS-Vector with a plaintext coefficient vector and a bias-addition.

### CODE REVIEW

File: he_man_tenseal/inference.py [hunk 1]
```diff
@@ -562,6 +563,9 @@ def execute(self, state: Dict[str, Union[ts.CKKSVector, np.ndarray]]) -> None:
     class AveragePoolOperator(GemmWrappedOperator):
         pass
 
+    class BatchNormalizationOperator(GemmWrappedOperator):
```
Comment (mrader1248) [inline he_man_tenseal/inference.py]: This is okay for the moment. I will open a PR to combine the previous version (fast, but only for plaintext parameters) with this version.

File: scripts/generate_mnist_calibration_data.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    test_loader = DataLoader(test_data, batch_size=1)
+    sample = next(iter(test_loader))[0].numpy().astype(np.float32)
+    # mnist_sample = test_dataset_array.reshape(10000, 1, 32, 32).astype(np.float32)[0]
+
+    path = Path(__file__).parent / ".." / "demo" / "mnist" / "input.npy"
+    np.save(path, sample)
+
+
+if __name__ == "__main__":
+    # generate_mnist_calibration_data()
```
Comment (mrader1248) [inline scripts/generate_mnist_calibration_data.py]: So, `generate_mnist_calibration_data` is never used?

File: scripts/batchnorm/batchnorm.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+
+if __name__ == "__main__":
+    # tests the BatchNormalizationOperator implementation in inference.py
+    #
+    # 2 models are generated, one for BatchNorm1d, one for BatchNorm2d.
+    # Models are trained 1 epoch with 1 batch to change initial BatchNorm parameters.
+    # Output for random inputs are computed and compared against HE-MAN output.
+    #
+    batchnorm_test()
```
Comment (mrader1248) [inline scripts/batchnorm/batchnorm.py:157]: Is this a regular test? If yes, why is it part of the helper scripts and not part of a unit test?

**Review by mrader1248 — COMMENTED** [review]
Thank you for this contribution and the necessary modifications for newer package versions.

My major concern is the slow-down of the `BatchNormalizationOperator` for plaintext paramters, i.e. `inputs[1:]`, but we will fix this with another PR.

File: scripts/generate_mnist_calibration_data.py (hunk 2 above)
Comment (MartinNoc) [reply scripts/generate_mnist_calibration_data.py]: Artifact of testing, will be resolved

File: scripts/batchnorm/batchnorm.py (hunk 3 above)
Comment (MartinNoc) [reply scripts/batchnorm/batchnorm.py:157]: unit tests (executed before a commit) already take so long - in my point of view this need not be tested for every commit

File: he_man_tenseal/inference.py (hunk 1 above)
Comment (MartinNoc) [reply he_man_tenseal/inference.py]: I will commit to this branch/PR

File: scripts/batchnorm/batchnorm.py (hunk 3 above)
Comment (mrader1248) [reply scripts/batchnorm/batchnorm.py:157]: I'm still not entirely comfortable with this, but let's keep it this way for the moment. Maybe we should think of a way to speed up the tests in general.

**Comment by mrader1248** [comment]
Thank you for this contribution. I really like the implementation of `execute` now. Please, feel free to merge if you have no more changes.

**Comment by MartinNoc** [comment]
@mrader1248 Thanks for reviewing!

**Comment by MartinNoc** [comment]
Had to do some fixes as I missed to test the ciphertext case. Hope it's okay to merge, we are in a bit of a hurry.
